QSP iterative angle solver

cvqmm opened the pull request:
**Context: QSP angle solver**

**Description of the Change: determine the QSP circuit angles through an iterative optimization procedure https://journals.aps.org/pra/abstract/10.1103/PhysRevA.103.042419**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
[[sc-91021](https://app.shortcut.com/xanaduai/story/91021)]

cvqmm commented:
@KetpuntoG @AntonNI8 @DSGuala

KetpuntoG commented:
Great! Thanks for working on it :)
Let me know if you have any question

cvqmm commented:
> Great! Thanks for working on it :) Let me know if you have any question

thx for the support.. just a couple:
- so you'll notice that I switched to random polynomials instead of hard coded coeffs for the tests.. the test that tests the complimentarity is falling short in terms of precision (~1e-5) instead (1e-7 which is the current atol) so idk how you/we want to resolve that (maybe reduce the atol)
- how to integrate jit: do I give the user the command to switch to jax use or do I do it under the hood by checking whether jax is installed and if it's the case switch from autograd to jax?

KetpuntoG commented:
> > Great! Thanks for working on it :) Let me know if you have any question
> 
> thx for the support.. just a couple:
> 
> * so you'll notice that I switched to random polynomials instead of hard coded coeffs for the tests.. the test that tests the complimentarity is falling short in terms of precision (~1e-5) instead (1e-7 which is the current atol) so idk how you/we want to resolve that (maybe reduce the atol)
> * how to integrate jit: do I give the user the command to switch to jax use or do I do it under the hood by checking whether jax is installed and if it's the case switch from autograd to jax?

- Let's reduce the atol to 1-e5 for the moment 👍
- Taking a look to the code, instead of using qml.numpy.____ you could use qml.math.___
In that way the interface will be chosen based on the input of the function

Don't worry about jit integration. The code itself will not have jit. But it would be good to create a test to check that it will indeed work with jit. [This is an example](https://github.com/PennyLaneAI/pennylane/blob/3935d4366d8f0a677b2bab482b9666d0585b7890/tests/templates/test_subroutines/test_reflection.py#L186) where jit is tested

cvqmm commented:
> > > Great! Thanks for working on it :) Let me know if you have any question
> > 
> > 
> > thx for the support.. just a couple:
> > 
> > * so you'll notice that I switched to random polynomials instead of hard coded coeffs for the tests.. the test that tests the complimentarity is falling short in terms of precision (~1e-5) instead (1e-7 which is the current atol) so idk how you/we want to resolve that (maybe reduce the atol)
> > * how to integrate jit: do I give the user the command to switch to jax use or do I do it under the hood by checking whether jax is installed and if it's the case switch from autograd to jax?
> 
> * Let's reduce the atol to 1-e5 for the moment 👍
> * Taking a look to the code, instead of using qml.numpy.____ you could use qml.math.___
>   In that way the interface will be chosen based on the input of the function
> 
> Don't worry about jit integration. The code itself will not have jit. But it would be good to create a test to check that it will indeed work with jit. [This is an example](https://github.com/PennyLaneAI/pennylane/blob/3935d4366d8f0a677b2bab482b9666d0585b7890/tests/templates/test_subroutines/test_reflection.py#L186) where jit is tested

- So I replaced qml.numpy with qml.math as suggested.. I also added a customed decorator that will jit the adequate functions if jax is available (which will also carry the test in jax if it's available).. let me know what you think..

DSGuala commented:
:rocket: Thanks for opening this PR @cvqmm !

AntonNI8 commented on tests/templates/test_subroutines/test_qsvt.py:
```diff
@@ -625,12 +625,35 @@ def test_global_phase_not_alway_applied():
 
 
 class TestRootFindingSolver:
```
@cvqmm I would actually recommend making a separate test class for the iterative angle solver! This way we can easily test polynomials that are suitable for the different solvers (ie. polynomials with degenerate roots).

KetpuntoG commented on pennylane/templates/subroutines/qsvt.py:
```diff
@@ -569,8 +571,154 @@ def _compute_qsp_angle(poly_coeffs):
+
+
+@jit_if_available
+def cheby_pol(x, degree):
```
Please see how the doc strings of other functions are doc strings to follow a similar style (“Args”, “Return”,...) 😄

KetpuntoG replied:
This applies also to the other functions

KetpuntoG reviewed (changes requested):
Thank you for working on this! Here are some comments :) 

On another note, I'm trying to use it and I can't manage to import pennylane from this branch `“AttributeError: partially initialized module ‘pennylane’ has no attribute ‘operation’ (most likely due to a circular import)”`
It is possible that removing some of the imports I mention in the comments will fix it. Make sure also that you resolve the conflicts with master, this could solve the problem

Let me know if something is not clear

cvqmm commented on pennylane/templates/subroutines/qsvt.py:
```diff
@@ -569,8 +571,154 @@ def _compute_qsp_angle(poly_coeffs):
+def _compute_qsp_angles_iteratively(polynomial_coeffs_in_cano_basis, opt_method="Newton-CG"):
+    polynomial_coeffs_in_cheby_basis = chebyshev.poly2cheb(polynomial_coeffs_in_cano_basis)
+    degree = len(polynomial_coeffs_in_cheby_basis) - 1
+    
+    coeffs_odd = polynomial_coeffs_in_cheby_basis[1::2]
+    coeffs_even = polynomial_coeffs_in_cheby_basis[0::2]
+
+    if np.allclose(coeffs_odd, np.zeros_like(coeffs_odd)):
+        coeffs_target_func = qml.math.array(coeffs_even)
+    elif np.allclose(coeffs_even, np.zeros_like(coeffs_even)):
+        coeffs_target_func = qml.math.array(coeffs_odd)
+    else:
+        raise ValueError()
```
the first check is done for the canonical basis.. this one is to make sure the Chebychev basis transformation hasn't screwed up the parity :) (it's more of a sanity check than user input check)

cvqmm commented on pennylane/templates/subroutines/qsvt.py:
```diff
@@ -569,8 +571,154 @@ def _compute_qsp_angle(poly_coeffs):
 
+import scipy
 
+def jit_if_available(func):
+    """
+    decorator that jax.jit the qsp_iterative optimization functions if jax is available
+    the static arguments to the jit function are meant to be contained in kwargs
+    """
+    try:
+        from functools import partial
```
sounds good!

cvqmm commented:
@KetpuntoG the reviews have been addressed.. ready for an extra round of reviews

KetpuntoG commented:
Thanks @cvqmm , I'll take a look 
One way to notify me in case you are waiting is to click on re-request in the top right, that way I won't miss it 😄

<img width="249" alt="Screenshot 2025-02-18 at 9 59 31 AM" src="https://github.com/user-attachments/assets/edcfeae6-3299-4bf3-9584-819da9d1cc14" />

KetpuntoG reviewed (commented):
Hi @cvqmm,

I have been testing the code and noticed that its performance is quite slow. Specifically, I ran an example using a polynomial of degree ~200, and the process was still running after two hours. Although it might seem that this issue is related to the removal of the JIT compilation component, I tested the [pyqsp](https://github.com/ichuang/pyqsp) library—which now incorporates a method based on the same paper—and obtained results in just five seconds without compilation 😮

Could you try to improve the efficiency of the current technique? I suspect that the bottleneck may be related to the use of SciPy, but it would be beneficial to verify this carefully.

Attached is the Chebyshev-based polynomial I have been testing, which approximates the function f(x) = 1/(100·x).

And thank you for the time spent on this 🙌

```python
poly = np.array([
    0, 0.01986268, 0., -0.01958816, 0., 0.0193139, 0., -0.01903993, 0.,
    0.01876648, 0., -0.01849345, 0., 0.01822035, 0., -0.01794791, 0.,
    0.01767489, 0., -0.01740275, 0., 0.01713172, 0., -0.01686185, 0.,
    0.01659323, 0., -0.01632495, 0., 0.01605625, 0., -0.01578922, 0.,
    0.01552611, 0., -0.01525896, 0., 0.01499665, 0., -0.01473404, 0.,
    0.01447287, 0., -0.01421316, 0., 0.01395486, 0., -0.013698, 0.,
    0.01344275, 0., -0.01318902, 0., 0.0129368, 0., -0.01268646, 0.,
    0.01243769, 0., -0.0121907, 0., 0.01194556, 0., -0.01170197, 0.,
    0.01146048, 0., -0.01122085, 0., 0.01098323, 0., -0.01074785, 0.,
    0.01051459, 0., -0.01028346, 0., 0.01005472, 0., -0.00982922, 0.,
    0.00959878, 0., -0.00938884, 0., 0.00916153, 0., -0.00894104, 0.,
    0.00872791, 0., -0.00851822, 0., 0.00830988, 0., -0.00810254, 0.,
    0.00789652, 0., -0.00769287, 0., 0.00749262, 0., -0.00729527, 0.,
    0.00710157, 0., -0.00691077, 0., 0.0067227, 0., -0.00653727, 0.,
    0.00635413, 0., -0.00617331, 0., 0.00599494, 0., -0.00581909, 0.,
    0.00564605, 0., -0.00547613, 0., 0.0053093, 0., -0.00514538, 0.,
    0.00498488, 0., -0.00482613, 0., 0.0046712, 0., -0.00451883, 0.,
    0.00436919, 0., -0.00422266, 0., 0.00407852, 0., -0.00393672, 0.,
    0.00379837, 0., -0.0036631, 0., 0.00352909, 0., -0.00339995, 0.,
    0.00327227, 0., -0.00314791, 0., 0.00302646, 0., -0.00290785, 0.,
    0.00279213, 0., -0.00267919, 0., 0.00256883, 0., -0.00246113, 0.,
    0.00235621, 0., -0.00225382, 0., 0.00215419, 0., -0.00205717, 0.,
    0.00196249, 0., -0.0018706, 0., 0.00178092, 0., -0.00169383, 0.,
    0.00160933, 0., -0.00152723, 0., 0.00144769, 0., -0.00137084, 0.,
    0.00129635, 0., -0.00122419, 0., 0.01031074
])
```

cvqmm commented:
> Hi @cvqmm,
> 
> I have been testing the code and noticed that its performance is quite slow. Specifically, I ran an example using a polynomial of degree ~200, and the process was still running after two hours. Although it might seem that this issue is related to the removal of the JIT compilation component, I tested the [pyqsp](https://github.com/ichuang/pyqsp) library—which now incorporates a method based on the same paper—and obtained results in just five seconds without compilation 😮
> 
> Could you try to improve the efficiency of the current technique? I suspect that the bottleneck may be related to the use of SciPy, but it would be beneficial to verify this carefully.
> 
> Attached is the Chebyshev-based polynomial I have been testing, which approximates the function f(x) = 1/(100·x).
> 
> And thank you for the time spent on this 🙌
> 
> ```python
> poly = np.array([
>     0, 0.01986268, 0., -0.01958816, 0., 0.0193139, 0., -0.01903993, 0.,
>     0.01876648, 0., -0.01849345, 0., 0.01822035, 0., -0.01794791, 0.,
>     0.01767489, 0., -0.01740275, 0., 0.01713172, 0., -0.01686185, 0.,
>     0.01659323, 0., -0.01632495, 0., 0.01605625, 0., -0.01578922, 0.,
>     0.01552611, 0., -0.01525896, 0., 0.01499665, 0., -0.01473404, 0.,
>     0.01447287, 0., -0.01421316, 0., 0.01395486, 0., -0.013698, 0.,
>     0.01344275, 0., -0.01318902, 0., 0.0129368, 0., -0.01268646, 0.,
>     0.01243769, 0., -0.0121907, 0., 0.01194556, 0., -0.01170197, 0.,
>     0.01146048, 0., -0.01122085, 0., 0.01098323, 0., -0.01074785, 0.,
>     0.01051459, 0., -0.01028346, 0., 0.01005472, 0., -0.00982922, 0.,
>     0.00959878, 0., -0.00938884, 0., 0.00916153, 0., -0.00894104, 0.,
>     0.00872791, 0., -0.00851822, 0., 0.00830988, 0., -0.00810254, 0.,
>     0.00789652, 0., -0.00769287, 0., 0.00749262, 0., -0.00729527, 0.,
>     0.00710157, 0., -0.00691077, 0., 0.0067227, 0., -0.00653727, 0.,
>     0.00635413, 0., -0.00617331, 0., 0.00599494, 0., -0.00581909, 0.,
>     0.00564605, 0., -0.00547613, 0., 0.0053093, 0., -0.00514538, 0.,
>     0.00498488, 0., -0.00482613, 0., 0.0046712, 0., -0.00451883, 0.,
>     0.00436919, 0., -0.00422266, 0., 0.00407852, 0., -0.00393672, 0.,
>     0.00379837, 0., -0.0036631, 0., 0.00352909, 0., -0.00339995, 0.,
>     0.00327227, 0., -0.00314791, 0., 0.00302646, 0., -0.00290785, 0.,
>     0.00279213, 0., -0.00267919, 0., 0.00256883, 0., -0.00246113, 0.,
>     0.00235621, 0., -0.00225382, 0., 0.00215419, 0., -0.00205717, 0.,
>     0.00196249, 0., -0.0018706, 0., 0.00178092, 0., -0.00169383, 0.,
>     0.00160933, 0., -0.00152723, 0., 0.00144769, 0., -0.00137084, 0.,
>     0.00129635, 0., -0.00122419, 0., 0.01031074
> ])
> ```

Hi @KetpuntoG, the bottleneck is coming from removing JIT compilation of JAX that accelerates the algebraic operations and not scipy. I briefly looked into pyqsp and it seems it also uses tensorflow Keras which (similarly to JAX) uses XLA for accelerating linear algebraic operations..   to summarize the the bottleneck is the qsp layers matrix product and not the optimization algorithm itself

KetpuntoG commented:
> > Hi @cvqmm,
> > I have been testing the code and noticed that its performance is quite slow. Specifically, I ran an example using a polynomial of degree ~200, and the process was still running after two hours. Although it might seem that this issue is related to the removal of the JIT compilation component, I tested the [pyqsp](https://github.com/ichuang/pyqsp) library—which now incorporates a method based on the same paper—and obtained results in just five seconds without compilation 😮
> > Could you try to improve the efficiency of the current technique? I suspect that the bottleneck may be related to the use of SciPy, but it would be beneficial to verify this carefully.
> > Attached is the Chebyshev-based polynomial I have been testing, which approximates the function f(x) = 1/(100·x).
> > And thank you for the time spent on this 🙌
> > ```python
> > poly = np.array([
> >     0, 0.01986268, 0., -0.01958816, 0., 0.0193139, 0., -0.01903993, 0.,
> >     0.01876648, 0., -0.01849345, 0., 0.01822035, 0., -0.01794791, 0.,
> >     0.01767489, 0., -0.01740275, 0., 0.01713172, 0., -0.01686185, 0.,
> >     0.01659323, 0., -0.01632495, 0., 0.01605625, 0., -0.01578922, 0.,
> >     0.01552611, 0., -0.01525896, 0., 0.01499665, 0., -0.01473404, 0.,
> >     0.01447287, 0., -0.01421316, 0., 0.01395486, 0., -0.013698, 0.,
> >     0.01344275, 0., -0.01318902, 0., 0.0129368, 0., -0.01268646, 0.,
> >     0.01243769, 0., -0.0121907, 0., 0.01194556, 0., -0.01170197, 0.,
> >     0.01146048, 0., -0.01122085, 0., 0.01098323, 0., -0.01074785, 0.,
> >     0.01051459, 0., -0.01028346, 0., 0.01005472, 0., -0.00982922, 0.,
> >     0.00959878, 0., -0.00938884, 0., 0.00916153, 0., -0.00894104, 0.,
> >     0.00872791, 0., -0.00851822, 0., 0.00830988, 0., -0.00810254, 0.,
> >     0.00789652, 0., -0.00769287, 0., 0.00749262, 0., -0.00729527, 0.,
> >     0.00710157, 0., -0.00691077, 0., 0.0067227, 0., -0.00653727, 0.,
> >     0.00635413, 0., -0.00617331, 0., 0.00599494, 0., -0.00581909, 0.,
> >     0.00564605, 0., -0.00547613, 0., 0.0053093, 0., -0.00514538, 0.,
> >     0.00498488, 0., -0.00482613, 0., 0.0046712, 0., -0.00451883, 0.,
> >     0.00436919, 0., -0.00422266, 0., 0.00407852, 0., -0.00393672, 0.,
> >     0.00379837, 0., -0.0036631, 0., 0.00352909, 0., -0.00339995, 0.,
> >     0.00327227, 0., -0.00314791, 0., 0.00302646, 0., -0.00290785, 0.,
> >     0.00279213, 0., -0.00267919, 0., 0.00256883, 0., -0.00246113, 0.,
> >     0.00235621, 0., -0.00225382, 0., 0.00215419, 0., -0.00205717, 0.,
> >     0.00196249, 0., -0.0018706, 0., 0.00178092, 0., -0.00169383, 0.,
> >     0.00160933, 0., -0.00152723, 0., 0.00144769, 0., -0.00137084, 0.,
> >     0.00129635, 0., -0.00122419, 0., 0.01031074
> > ])
> > ```
> 
> Hi @KetpuntoG, the bottleneck is coming from removing JIT compilation of JAX that accelerates the algebraic operations and not scipy. I briefly looked into pyqsp and it seems it also uses tensorflow Keras which (similarly to JAX) uses XLA for accelerating linear algebraic operations.. to summarize the the bottleneck is the qsp layers matrix product and not the optimization algorithm itself

Hi! I've been talking with the team and an approach that it is valid with jax in pennylane would be something like this 😄

```
        elif angle_solver == "iterative":
            try:
                import jax
                poly = jax.numpy.array(poly)
                return jax.jit(_compute_qsp_angles_iteratively)(poly)
```

Do you get the expected performance with this?

cvqmm commented:
> > > Hi @cvqmm,
> > > I have been testing the code and noticed that its performance is quite slow. Specifically, I ran an example using a polynomial of degree ~200, and the process was still running after two hours. Although it might seem that this issue is related to the removal of the JIT compilation component, I tested the [pyqsp](https://github.com/ichuang/pyqsp) library—which now incorporates a method based on the same paper—and obtained results in just five seconds without compilation 😮
> > > Could you try to improve the efficiency of the current technique? I suspect that the bottleneck may be related to the use of SciPy, but it would be beneficial to verify this carefully.
> > > Attached is the Chebyshev-based polynomial I have been testing, which approximates the function f(x) = 1/(100·x).
> > > And thank you for the time spent on this 🙌
> > > ```python
> > > poly = np.array([
> > >     0, 0.01986268, 0., -0.01958816, 0., 0.0193139, 0., -0.01903993, 0.,
> > >     0.01876648, 0., -0.01849345, 0., 0.01822035, 0., -0.01794791, 0.,
> > >     0.01767489, 0., -0.01740275, 0., 0.01713172, 0., -0.01686185, 0.,
> > >     0.01659323, 0., -0.01632495, 0., 0.01605625, 0., -0.01578922, 0.,
> > >     0.01552611, 0., -0.01525896, 0., 0.01499665, 0., -0.01473404, 0.,
> > >     0.01447287, 0., -0.01421316, 0., 0.01395486, 0., -0.013698, 0.,
> > >     0.01344275, 0., -0.01318902, 0., 0.0129368, 0., -0.01268646, 0.,
> > >     0.01243769, 0., -0.0121907, 0., 0.01194556, 0., -0.01170197, 0.,
> > >     0.01146048, 0., -0.01122085, 0., 0.01098323, 0., -0.01074785, 0.,
> > >     0.01051459, 0., -0.01028346, 0., 0.01005472, 0., -0.00982922, 0.,
> > >     0.00959878, 0., -0.00938884, 0., 0.00916153, 0., -0.00894104, 0.,
> > >     0.00872791, 0., -0.00851822, 0., 0.00830988, 0., -0.00810254, 0.,
> > >     0.00789652, 0., -0.00769287, 0., 0.00749262, 0., -0.00729527, 0.,
> > >     0.00710157, 0., -0.00691077, 0., 0.0067227, 0., -0.00653727, 0.,
> > >     0.00635413, 0., -0.00617331, 0., 0.00599494, 0., -0.00581909, 0.,
> > >     0.00564605, 0., -0.00547613, 0., 0.0053093, 0., -0.00514538, 0.,
> > >     0.00498488, 0., -0.00482613, 0., 0.0046712, 0., -0.00451883, 0.,
> > >     0.00436919, 0., -0.00422266, 0., 0.00407852, 0., -0.00393672, 0.,
> > >     0.00379837, 0., -0.0036631, 0., 0.00352909, 0., -0.00339995, 0.,
> > >     0.00327227, 0., -0.00314791, 0., 0.00302646, 0., -0.00290785, 0.,
> > >     0.00279213, 0., -0.00267919, 0., 0.00256883, 0., -0.00246113, 0.,
> > >     0.00235621, 0., -0.00225382, 0., 0.00215419, 0., -0.00205717, 0.,
> > >     0.00196249, 0., -0.0018706, 0., 0.00178092, 0., -0.00169383, 0.,
> > >     0.00160933, 0., -0.00152723, 0., 0.00144769, 0., -0.00137084, 0.,
> > >     0.00129635, 0., -0.00122419, 0., 0.01031074
> > > ])
> > > ```
> > 
> > 
> > Hi @KetpuntoG, the bottleneck is coming from removing JIT compilation of JAX that accelerates the algebraic operations and not scipy. I briefly looked into pyqsp and it seems it also uses tensorflow Keras which (similarly to JAX) uses XLA for accelerating linear algebraic operations.. to summarize the the bottleneck is the qsp layers matrix product and not the optimization algorithm itself
> 
> Hi! I've been talking with the team and an approach that it is valid with jax in pennylane would be something like this 😄
> 
> ```
>         elif angle_solver == "iterative":
>             try:
>                 import jax
>                 poly = jax.numpy.array(poly)
>                 return jax.jit(_compute_qsp_angles_iteratively)(poly)
>             except:
>                 return _compute_qsp_angles_iteratively(poly)
> ```
> 
> Do you get the expected performance with this?

Hi :) so the _compute_qsp_angles_iteratively is not JIT compilable because the scipy optimizers are not JIT compilable.. but the performance comes from JIT compiling the cost function.. so that needs to be done one level up i.e. is inside the _compute_qsp_angles_iteratively function.. I can do that if that's OK?

KetpuntoG commented:
That sounds great @cvqmm 👌
The important thing is to make the `import` inside the `try`

KetpuntoG commented:
Hello @cvqmm, thank you for this update!
I have tested it with polynomials of degree 200 and now the time has been significantly reduced to 2 minutes 🚀

I decided to push it a bit further  and when running a polynomial of degree 1000, the performance was reduced considerably (>>2h). Surprisedly, the same method with `pyqsp`, it remained under two minutes which makes me think that we can still do better in the current implementation 💪

First of all, I'd like to share with you the script I used just in case we are accidentally using different versions:

```
import pennylane as qml
import numpy as np
import time

time0 = time.time()

poly = []

degree = 1000

# we generate a random poly
for i in range(degree //2):
  poly.append(0)
  poly.append(np.random.rand() / (degree // 2))

qsvt_angles = qml.poly_to_angles(poly, "QSVT", angle_solver = "iterative")

x = 0.2

# Encode x in the top left of the matrix
block_encoding = qml.RX(2 * np.arccos(x), wires=0)
projectors = [qml.PCPhase(angle, dim=1, wires=0) for angle in qsvt_angles]

@qml.qnode(qml.device("default.qubit"))
def circuit_qsvt():
    qml.QSVT(block_encoding, projectors)
    return qml.state()

output = qml.matrix(circuit_qsvt, wire_order=[0])()[0, 0]
expected = sum(coef * (x**i) for i, coef in enumerate(poly))

print("output qsvt: ", output.real)
print("P(x) =       ", expected)

print(time.time() - time0)
```

Could you confirm that the performance I get is the same as the performance you get? 
And thanks again for your time, it is not an easy challenge 🙌

cvqmm commented:
> Hello @cvqmm, thank you for this update! I have tested it with polynomials of degree 200 and now the time has been significantly reduced to 2 minutes 🚀
> 
> I decided to push it a bit further and when running a polynomial of degree 1000, the performance was reduced considerably (>>2h). Surprisedly, the same method with `pyqsp`, it remained under two minutes which makes me think that we can still do better in the current implementation 💪
> 
> First of all, I'd like to share with you the script I used just in case we are accidentally using different versions:
> 
> ```
> import pennylane as qml
> import numpy as np
> import time
> 
> time0 = time.time()
> 
> poly = []
> 
> degree = 1000
> 
> # we generate a random poly
> for i in range(degree //2):
>   poly.append(0)
>   poly.append(np.random.rand() / (degree // 2))
> 
> qsvt_angles = qml.poly_to_angles(poly, "QSVT", angle_solver = "iterative")
> 
> x = 0.2
> 
> # Encode x in the top left of the matrix
> block_encoding = qml.RX(2 * np.arccos(x), wires=0)
> projectors = [qml.PCPhase(angle, dim=1, wires=0) for angle in qsvt_angles]
> 
> @qml.qnode(qml.device("default.qubit"))
> def circuit_qsvt():
>     qml.QSVT(block_encoding, projectors)
>     return qml.state()
> 
> output = qml.matrix(circuit_qsvt, wire_order=[0])()[0, 0]
> expected = sum(coef * (x**i) for i, coef in enumerate(poly))
> 
> print("output qsvt: ", output.real)
> print("P(x) =       ", expected)
> 
> print(time.time() - time0)
> ```
> 
> Could you confirm that the performance I get is the same as the performance you get? And thanks again for your time, it is not an easy challenge 🙌

Hi @KetpuntoG! no worries it's actually good to have something to benchmark against.. so I managed to further optimize the code and bring down the time for degree 1000 to under a minute.. I'll push it once cleaned.

cvqmm commented:
> > Hello @cvqmm, thank you for this update! I have tested it with polynomials of degree 200 and now the time has been significantly reduced to 2 minutes 🚀
> > I decided to push it a bit further and when running a polynomial of degree 1000, the performance was reduced considerably (>>2h). Surprisedly, the same method with `pyqsp`, it remained under two minutes which makes me think that we can still do better in the current implementation 💪
> > First of all, I'd like to share with you the script I used just in case we are accidentally using different versions:
> > ```
> > import pennylane as qml
> > import numpy as np
> > import time
> > 
> > time0 = time.time()
> > 
> > poly = []
> > 
> > degree = 1000
> > 
> > # we generate a random poly
> > for i in range(degree //2):
> >   poly.append(0)
> >   poly.append(np.random.rand() / (degree // 2))
> > 
> > qsvt_angles = qml.poly_to_angles(poly, "QSVT", angle_solver = "iterative")
> > 
> > x = 0.2
> > 
> > # Encode x in the top left of the matrix
> > block_encoding = qml.RX(2 * np.arccos(x), wires=0)
> > projectors = [qml.PCPhase(angle, dim=1, wires=0) for angle in qsvt_angles]
> > 
> > @qml.qnode(qml.device("default.qubit"))
> > def circuit_qsvt():
> >     qml.QSVT(block_encoding, projectors)
> >     return qml.state()
> > 
> > output = qml.matrix(circuit_qsvt, wire_order=[0])()[0, 0]
> > expected = sum(coef * (x**i) for i, coef in enumerate(poly))
> > 
> > print("output qsvt: ", output.real)
> > print("P(x) =       ", expected)
> > 
> > print(time.time() - time0)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Could you confirm that the performance I get is the same as the performance you get? And thanks again for your time, it is not an easy challenge 🙌
> 
> Hi @KetpuntoG! no worries it's actually good to have something to benchmark against.. so I managed to further optimize the code and bring down the time for degree 1000 to under a minute.. I'll push it once cleaned.

@KetpuntoG have u tried running your test case? let me know what you get..

cvqmm commented:
> > > > Hello @cvqmm, thank you for this update! I have tested it with polynomials of degree 200 and now the time has been significantly reduced to 2 minutes 🚀
> > > > I decided to push it a bit further and when running a polynomial of degree 1000, the performance was reduced considerably (>>2h). Surprisedly, the same method with `pyqsp`, it remained under two minutes which makes me think that we can still do better in the current implementation 💪
> > > > First of all, I'd like to share with you the script I used just in case we are accidentally using different versions:
> > > > ```
> > > > import pennylane as qml
> > > > import numpy as np
> > > > import time
> > > > 
> > > > time0 = time.time()
> > > > 
> > > > poly = []
> > > > 
> > > > degree = 1000
> > > > 
> > > > # we generate a random poly
> > > > for i in range(degree //2):
> > > >   poly.append(0)
> > > >   poly.append(np.random.rand() / (degree // 2))
> > > > 
> > > > qsvt_angles = qml.poly_to_angles(poly, "QSVT", angle_solver = "iterative")
> > > > 
> > > > x = 0.2
> > > > 
> > > > # Encode x in the top left of the matrix
> > > > block_encoding = qml.RX(2 * np.arccos(x), wires=0)
> > > > projectors = [qml.PCPhase(angle, dim=1, wires=0) for angle in qsvt_angles]
> > > > 
> > > > @qml.qnode(qml.device("default.qubit"))
> > > > def circuit_qsvt():
> > > >     qml.QSVT(block_encoding, projectors)
> > > >     return qml.state()
> > > > 
> > > > output = qml.matrix(circuit_qsvt, wire_order=[0])()[0, 0]
> > > > expected = sum(coef * (x**i) for i, coef in enumerate(poly))
> > > > 
> > > > print("output qsvt: ", output.real)
> > > > print("P(x) =       ", expected)
> > > > 
> > > > print(time.time() - time0)
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Could you confirm that the performance I get is the same as the performance you get? And thanks again for your time, it is not an easy challenge 🙌
> > > 
> > > 
> > > Hi @KetpuntoG! no worries it's actually good to have something to benchmark against.. so I managed to further optimize the code and bring down the time for degree 1000 to under a minute.. I'll push it once cleaned.
> > 
> > 
> > @KetpuntoG have u tried running your test case? let me know what you get..
> 
> Hi @cvqmm , It still takes >> 2h in the case of degree 1000. I have double checked and I am using the current version. My computer is a mac M1, what do you think our difference in performance is due to?

Maybe check your JAX installation (whether JAX version is compatible with the CUDA version).. I'm running on a linux node with 8 cpu cores which is similar to ur specs I think.. also I'm running JAX on cpu only which theoretically means it should be faster with gpu

KetpuntoG commented:
> > > > > Hello @cvqmm, thank you for this update! I have tested it with polynomials of degree 200 and now the time has been significantly reduced to 2 minutes 🚀
> > > > > I decided to push it a bit further and when running a polynomial of degree 1000, the performance was reduced considerably (>>2h). Surprisedly, the same method with `pyqsp`, it remained under two minutes which makes me think that we can still do better in the current implementation 💪
> > > > > First of all, I'd like to share with you the script I used just in case we are accidentally using different versions:
> > > > > ```
> > > > > import pennylane as qml
> > > > > import numpy as np
> > > > > import time
> > > > > 
> > > > > time0 = time.time()
> > > > > 
> > > > > poly = []
> > > > > 
> > > > > degree = 1000
> > > > > 
> > > > > # we generate a random poly
> > > > > for i in range(degree //2):
> > > > >   poly.append(0)
> > > > >   poly.append(np.random.rand() / (degree // 2))
> > > > > 
> > > > > qsvt_angles = qml.poly_to_angles(poly, "QSVT", angle_solver = "iterative")
> > > > > 
> > > > > x = 0.2
> > > > > 
> > > > > # Encode x in the top left of the matrix
> > > > > block_encoding = qml.RX(2 * np.arccos(x), wires=0)
> > > > > projectors = [qml.PCPhase(angle, dim=1, wires=0) for angle in qsvt_angles]
> > > > > 
> > > > > @qml.qnode(qml.device("default.qubit"))
> > > > > def circuit_qsvt():
> > > > >     qml.QSVT(block_encoding, projectors)
> > > > >     return qml.state()
> > > > > 
> > > > > output = qml.matrix(circuit_qsvt, wire_order=[0])()[0, 0]
> > > > > expected = sum(coef * (x**i) for i, coef in enumerate(poly))
> > > > > 
> > > > > print("output qsvt: ", output.real)
> > > > > print("P(x) =       ", expected)
> > > > > 
> > > > > print(time.time() - time0)
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > Could you confirm that the performance I get is the same as the performance you get? And thanks again for your time, it is not an easy challenge 🙌
> > > > 
> > > > 
> > > > Hi @KetpuntoG! no worries it's actually good to have something to benchmark against.. so I managed to further optimize the code and bring down the time for degree 1000 to under a minute.. I'll push it once cleaned.
> > > 
> > > 
> > > @KetpuntoG have u tried running your test case? let me know what you get..
> > 
> > 
> > Hi @cvqmm , It still takes >> 2h in the case of degree 1000. I have double checked and I am using the current version. My computer is a mac M1, what do you think our difference in performance is due to?
> 
> Maybe check your JAX installation (whether JAX version is compatible with the CUDA version).. I'm running on a linux node with 8 cpu cores which is similar to ur specs I think.. also I'm running JAX on cpu only which theoretically means it should be faster with gpu

I can't quite get it yet 🤔 I also tried Google Colab in GPU. Do you think you could check if google colab works for you? This way you could share me a link and I could work there. I am installing your branch with the command
` !pip install git+https://github.com/cvqmm/pennylane.git@qsp_iterative`

cvqmm commented:
> > > > > > Hello @cvqmm, thank you for this update! I have tested it with polynomials of degree 200 and now the time has been significantly reduced to 2 minutes 🚀
> > > > > > I decided to push it a bit further and when running a polynomial of degree 1000, the performance was reduced considerably (>>2h). Surprisedly, the same method with `pyqsp`, it remained under two minutes which makes me think that we can still do better in the current implementation 💪
> > > > > > First of all, I'd like to share with you the script I used just in case we are accidentally using different versions:
> > > > > > ```
> > > > > > import pennylane as qml
> > > > > > import numpy as np
> > > > > > import time
> > > > > > 
> > > > > > time0 = time.time()
> > > > > > 
> > > > > > poly = []
> > > > > > 
> > > > > > degree = 1000
> > > > > > 
> > > > > > # we generate a random poly
> > > > > > for i in range(degree //2):
> > > > > >   poly.append(0)
> > > > > >   poly.append(np.random.rand() / (degree // 2))
> > > > > > 
> > > > > > qsvt_angles = qml.poly_to_angles(poly, "QSVT", angle_solver = "iterative")
> > > > > > 
> > > > > > x = 0.2
> > > > > > 
> > > > > > # Encode x in the top left of the matrix
> > > > > > block_encoding = qml.RX(2 * np.arccos(x), wires=0)
> > > > > > projectors = [qml.PCPhase(angle, dim=1, wires=0) for angle in qsvt_angles]
> > > > > > 
> > > > > > @qml.qnode(qml.device("default.qubit"))
> > > > > > def circuit_qsvt():
> > > > > >     qml.QSVT(block_encoding, projectors)
> > > > > >     return qml.state()
> > > > > > 
> > > > > > output = qml.matrix(circuit_qsvt, wire_order=[0])()[0, 0]
> > > > > > expected = sum(coef * (x**i) for i, coef in enumerate(poly))
> > > > > > 
> > > > > > print("output qsvt: ", output.real)
> > > > > > print("P(x) =       ", expected)
> > > > > > 
> > > > > > print(time.time() - time0)
> > > > > > ```
> > > > > > 
> > > > > > 
> > > > > >     
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >     
> > > > > >   
> > > > > > Could you confirm that the performance I get is the same as the performance you get? And thanks again for your time, it is not an easy challenge 🙌
> > > > > 
> > > > > 
> > > > > Hi @KetpuntoG! no worries it's actually good to have something to benchmark against.. so I managed to further optimize the code and bring down the time for degree 1000 to under a minute.. I'll push it once cleaned.
> > > > 
> > > > 
> > > > @KetpuntoG have u tried running your test case? let me know what you get..
> > > 
> > > 
> > > Hi @cvqmm , It still takes >> 2h in the case of degree 1000. I have double checked and I am using the current version. My computer is a mac M1, what do you think our difference in performance is due to?
> > 
> > 
> > Maybe check your JAX installation (whether JAX version is compatible with the CUDA version).. I'm running on a linux node with 8 cpu cores which is similar to ur specs I think.. also I'm running JAX on cpu only which theoretically means it should be faster with gpu
> 
> I can't quite get it yet 🤔 I also tried Google Colab in GPU. Do you think you could check if google colab works for you? This way you could share me a link and I could work there. I am installing your branch with the command ` !pip install git+https://github.com/cvqmm/pennylane.git@qsp_iterative`

just tried colab and it's working fine.. send me your email so I can share it w/ u

KetpuntoG commented:
Hi @cvqmm 😄  Just a quick note to let you know that I’m taking a look to the PR and will be able to upload the review later this week

KetpuntoG reviewed (changes requested):
The PR looks great! Definitely a feature that I will use a lot 😄
I’ve left some comments. I’m happy to work on some of these suggestions, so please let us know if you’d like us to take over some of the work.
Thank you for contributing this feature!

cvqmm commented on tests/templates/test_subroutines/test_qsvt.py:
```diff
@@ -22,7 +22,7 @@
 import pennylane as qml
 from pennylane import numpy as np
-from pennylane.templates.subroutines.qsvt import _complementary_poly
+from pennylane.templates.subroutines.qsvt import _complementary_poly, cheby_pol, poly_func, qsp_iterates, qsp_optimization
```
I've added tests for all the intermediate functions except for qsp_optimization which is already covered by `test_correctness_QSP_angles_finding` and `test_correctness_QSVT_angles`..

KetpuntoG reviewed (commented):
Thank you very much for these recent changes. The PR is now almost complete 🙌 I will take it from here and do a few final minor adjustments. I would like to list you as a contributor for this release, if that is okey with you 😄 If so, could you please provide your full name so I can include it in the changelog? Excellent work once again!

AntonNI8 commented on doc/releases/changelog-dev.md:
```diff
@@ -12,6 +12,10 @@
   gate-set that can be translated to the MBQC formalism.
   [(7271)](https://github.com/PennyLaneAI/pennylane/pull/7271)
 
+* a new `angle_solver` method has been added in the `poly_to_angles` function. This allows to 
+  calculate the angles of polynomial degrees greater than our current method.
+ [(6694)](https://github.com/PennyLaneAI/pennylane/pull/6694)
+* 
```
Just the changelog 🎯 
```suggestion
* A new iterative angle solver for QSVT and QSP is available in the `poly_to_angles` function,
  allowing angle computation for polynomials of large degrees (>1000).
  Set `angle_solver="iterative"` in `poly_to_angles`, or from `qsvt` to use it.
  [(6694)](https://github.com/PennyLaneAI/pennylane/pull/6694)

```

JerryChen97 commented:
@KetpuntoG Mind having a story for this PR?

AntonNI8 commented:
[sc-91021] @JerryChen97

JerryChen97 reviewed (commented):
Thanks a lot for all the effort to introduce the new solver! Haven't got time to explore more the original literature; here are some general comments from dev/maintenance perspective

cvqmm commented on pennylane/templates/subroutines/qsvt.py:
```diff
@@ -753,6 +768,223 @@ def _compute_qsp_angle(poly_coeffs):
+    return phis, cost_func
+
+
+def _compute_qsp_angles_iteratively(poly):
```
it calls `qsp_optimization` which is iterative..

cvqmm commented on pennylane/templates/subroutines/qsvt.py:
```diff
@@ -753,6 +768,223 @@ def _compute_qsp_angle(poly_coeffs):
+    coeffs_even = poly_cheb[0::2]
+
+    if np.allclose(coeffs_odd, np.zeros_like(coeffs_odd)):
+        coeffs_target_func = qml.math.array(coeffs_even)
```
That should never happen as this method would be accessed by the user through the function `poly_to_angles` which handles the all zero coeffs case.. and if someone is experienced/familiar enough with the code to directly invoke this function then the null polynomial function is an "obvious" edge and "uninteresting" case :) .. @KetpuntoG plz correct me if I'm wrong

KetpuntoG commented on tests/templates/test_subroutines/test_qsvt.py:
```diff
@@ -40,6 +50,32 @@ def lst_phis(phis):
 
 
+def generate_polynomial_coeffs(degree, parity=None):
+    rng = np.random.default_rng(seed=123)
```
It should work for arbitrary polynomials, we have chosen a seed simply to have fixed tests 😄

JerryChen97 commented on pennylane/templates/subroutines/qsvt.py:
```diff
@@ -1003,11 +1240,16 @@ def circuit_qsvt():
         if angle_solver == "root-finding":
             return transform_angles(_compute_qsp_angle(poly), "QSP", "QSVT")
+        if angle_solver == "iterative":
+            return transform_angles(_compute_qsp_angles_iteratively(poly), "QSP", "QSVT")
```
```suggestion
            return transform_angles(
        f"Invalid angle_solver '{angle_solver}' for QSVT. Supported solvers are 'root-finding' and 'iterative'."
    )
```

JerryChen97 commented on pennylane/templates/subroutines/qsvt.py:
```diff
@@ -1003,11 +1240,16 @@ def circuit_qsvt():
+            return _compute_qsp_angles_iteratively(poly)
         raise AssertionError("Invalid angle solver method. Valid value is 'root-finding'")
 
     if routine == "GQSP":
```
Mind also help correct the next assertion error message below? I can't make comment on webui of gh

JerryChen97 commented on pennylane/templates/subroutines/qsvt.py:
```diff
@@ -753,6 +768,225 @@ def _compute_qsp_angle(poly_coeffs):
+    """
+
+    ind = qml.math.arange(len(coeffs))
+    return sum([coeffs[i] * _cheby_pol(x, degree=2 * i + parity) for i in ind])
```
```suggestion
    return sum(coeffs[i] * _cheby_pol(x, degree=2 * i + parity) for i in ind)
```

KetpuntoG reviewed (approved):
🚀

AntonNI8 commented on pennylane/templates/subroutines/qsvt.py:
```diff
@@ -918,10 +1152,13 @@ def poly_to_angles(poly, routine, angle_solver: Literal["root-finding"] = "root-
 
-        angle_solver (str): the method used to calculate the angles. Default is ``"root-finding"``.
-            ``"root-finding"`` is currently the only supported method.
+        angle_solver (str): The method used to calculate the angles; either ``"root-finding"`` or ``"iterative"``.
+            Default is ``"root-finding"``. ``"root-finding"`` is a method that works with all three routines, and
+            is effective for polynomials of degree up to :math:`\sim 1000`.
+            ``"iterative"`` employs an optimization method allowing angle computation
+            for polynomials of higher degree (:math:`\gt 1000`) for the ``"QSP"`` and ``"QSVT"`` routines.
```
```suggestion
        angle_solver (str): The method used to calculate the angles; either ``"root-finding"`` or ``"iterative"``.
            Default is ``"root-finding"``.
            
            - ``"root-finding"``: a method that works with all three routines, and
            is effective for polynomials of degree up to :math:`\sim 1000`.
            - ``"iterative"``: employs an optimization method allowing angle computation
            for polynomials of higher degree (:math:`\gt 1000`) for the ``"QSP"`` and ``"QSVT"`` routines.
```

AntonNI8 commented on doc/releases/changelog-dev.md:
```diff
@@ -68,6 +68,12 @@
   gate-set that can be translated to the MBQC formalism.
   [(7271)](https://github.com/PennyLaneAI/pennylane/pull/7271)
 
+* A new iterative angle solver for QSVT and QSP is available in the :func:`poly_to_angles <pennylane.poly_to_angles>` function,
+  allowing angle computation for polynomials of large degrees (>1000).
+  Set `angle_solver="iterative"` in the :func:`poly_to_angles  <pennylane.poly_to_angles>` function
+  (or from the :func:`qsvt <pennylane.qsvt>` function!) to use it.
+  [(6694)](https://github.com/PennyLaneAI/pennylane/pull/6694)
```
```suggestion
* A new iterative angle solver for QSVT and QSP is available in the :func:`poly_to_angles <pennylane.poly_to_angles>` function,
  allowing angle computation for polynomials of large degrees (:math:`\gt 1000`).
  Set `angle_solver="iterative"` in the :func:`poly_to_angles  <pennylane.poly_to_angles>` function
  (or from the :func:`qsvt <pennylane.qsvt>` function!) to use it.
  [(6694)](https://github.com/PennyLaneAI/pennylane/pull/6694)
```

AntonNI8 reviewed (approved):
Awesome 🚀 just some minor comments.
